geps/gep-1731/index.md

```diff
@@ -0,0 +1,436 @@
+    // still in progress when the Request timeout duration has been reached,
+    // these SHOULD be cancelled if possible and the Gateway MUST immediately
+    // return a timeout error.
+    //
+    // If a BackendRequest timeout (`rules[].timeouts.backendRequest`) is
+    // configured on the route, any retry attempts which reach the configured
+    // BackendRequest timeout duration without a response SHOULD be cancelled if
+    // possible and the Gateway should wait for at least the specified backoff
+    // duration before attempting to retry the backend request again.
+    //
+    // If a BackendRequest timeout is _not_ configured on the route, retry
+    // attempts MAY time out after an implementation default duration, or MAY
+    // remain pending until a configured Request timeout or implementation
+    // default duration for total request time is reached.
+    //
+    // When this field is unspecified, the time to wait between retry attempts
+    // is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Backoff *Duration `json:"backoff,omitempty"`
+}
+
+// HTTPRouteRetryStatusCode defines an HTTP response status code for
+// which a backend request should be retried.
+//
+// Implementations MUST support the following status codes as retryable:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// Implementations SHOULD support the `"5xx"` shorthand for matching status
+// codes in the 500-599 range. For implementations supporting `"5xx"`, this
+// shorthand MUST match the following status codes:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// The `"5xx"` shorthand MAY additionally match a well-documented subset of
+// arbitrary status codes in the 500-599 range, or match all status codes in
+// this range.
+//
+// Implementations SHOULD NOT support retrying status codes in the 100-399
+// range, as these responses are generally not appropriate to retry.
+//
+// Implementations MAY support specifying discrete values in the 400-499 range,
+// which are often inadvisable to retry, and MAY support discrete values in the
+// 600-999 (inclusive) range, which are not valid for HTTP clients, but are
+// sometimes used for communicating application-specific errors.
+//
+// +kubebuilder:validation:Pattern=`^[1-9](?:[0-9][0-9]|xx)$`
+type HTTPRouteRetryStatusCode string
+
+// Duration is a string value representing a duration in time. The foramat is
+// as specified in GEP-2257, a strict subset of the syntax parsed by Golang
+// time.ParseDuration.
+//
+// +kubebuilder:validation:Pattern=`^([0-9]{1,5}(h|m|s|ms)){1,4}$`
+type Duration string
+```
+
+### YAML
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: HTTPRoute
+metadata:
+  name: retry-example
+spec:
+  ...
+  rules:
+  - backendRefs:
+    - name: some-service
+      port: 8080
+    retry:
+      codes:
+      - 5xx
+      attempts: 2
+      backoff: 100ms
```

Comment:
This is a nice example showing retry for a backend service.  Would be good to have an example of retry for the entire route as well, which I assume it is possible.

Reply:
Hmm, that's a good question and **I think the YAML formatting is actually indented slightly wrong here, giving an inaccurate impression of the proposed structure** - I'll fix that.

---

This is currently proposed as scoping the `retry` stanza to an individual HTTPRouteRule (and _all_ backendRefs on the rule), which is the same scoping as [`timeouts`](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteTimeouts), [`filters`](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteFilter) and [`sessionPersistance`](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.SessionPersistence).

Given that the `matches` clauses in a single rule are OR'd together, a single HTTPRouteRule clause can be quite comprehensive, and would generally be sufficient for configuring the retry logic for many inbound paths/matches to a given set of backends (such as `foo` and a `foo-v2` canary).
> Each match is independent, i.e. this rule will be matched if any one of the matches is satisfied.
https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.HTTPRouteRule

We could _additionally_ allow specifying a more granular retry stanza for a specific backend within a rule (which I believe is what's currently implied by the inaccurate YAML), which has prior art with [`filters`](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteFilter), but I don't think that's necessary in an initial implementation and should only be considered if we get significant user feedback asking for that capability.

Reply:
FWIW I agree that per-backend retries is a thing we should only do with strong user feedback. If the common case is for Ana to consider all the backends on a route as logically equivalent (which I believe it is) then I expect that we won't need per-backend retries for a long while.